### ISSUE TITLE
Compiler does not report unused imports when XML doc comments are disabled

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#41640]
https://github.com/dotnet/roslyn/pull/41363 moves the "Remove unnecessary usings" IDE analyzer into the CodeStyle NuGet package, so this analyzer can execute on CI. This analyzer depends on the compiler layer reporting hidden diagnostics (`CS8019`) for unused usings: https://github.com/dotnet/roslyn/blob/2b12b9f3b8fe8895694280648b523bc4fd95211f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Helpers/RemoveUnnecessaryImports/CSharpUnnecessaryImportsProvider.cs#L23-L48.

However, the logic in compiler layer that reports these unused usings is guarded on XML doc comments being disabled: https://github.com/dotnet/roslyn/blob/41bc8954aa449ee96a0858de5f614f8a61fa2cad/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L2116-L2140

This logic seems to work fine for IDE-only purpose as XML doc comments are explicitly force enabled in the IDE. However, it breaks down for this same analyzer from command line when XML doc comments are disabled.

I presume this was done as a performance optimization to prevent this code from executing on default command line builds, when XML doc comments are disabled. I see 2 options to retain this performance optimization while fixing the command line case:

1. Expose a new public API on Compilation or SemanticModel to for unused usings. The existing compiler code can invoke the same functionality wrapped around the XML documentation mode check, and we can move the IDE's unnecessary usings analyzer to use this new API instead of `CS8019` diagnostics.
2. Loosen the compiler layer's guard around the above code to also execute this code when the effective severity of IDE0005 (unnecessary usings IDE diagnostic) is warning or above. This seems more simpler, but also more fragile change.

IMO, option 1. is a much better approach.

**Comment by mavasani** [comment]
Tagging @jcouv @sharwell as they seem to have good context on this IDE analyzer by looking at past issues.

**Comment by CyrusNajmabadi** [comment]
I thought we had a flag here to at least tell the compiler to bind the docs, even if we don't want errors or xml generation for them.  That way our symbolic information is all correct, regardless of what emit behavior the user specifies.

**Comment by sharwell** [comment]
@mavasani The underlying compiler diagnostic (CS8019) is not accurate if XML documentation parsing is disabled. Currently the project can only specify `DocumentationMode.None` or `DocumentationMode.Diagnose`. To enable IDE0005 in command line scenarios _without_ enabling XML documentation file output, we need a way to set the documentation mode to `DocumentationMode.Parse`.

**Comment by mavasani** [comment]
Ah, maybe in that case we can enhance the unused usings analyzer to report a separate diagnostic when IDE0005 is bumped to warning/error but documentation mode is None, which informs the user that XML documentation comments need to be enabled for the project to get accurate IDE0005 on CI. Does that sound reasonable? If so, we can move this issue to Area-IDE.

**Comment by jnm2** [comment]
For people wondering why command-line/CI builds are letting unused usings through, this is what you need to paste into your csproj or Directory.Build.props after adding `<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>` and editorconfig `dotnet_diagnostic.IDE0005.severity = warning`:

```xml
    <!-- Workaround for https://github.com/dotnet/roslyn/issues/53720 -->
    <PropertyGroup>
      <GenerateDocumentationFile>true</GenerateDocumentationFile>
      <NoWarn>$(NoWarn);CS1591;CS1573</NoWarn>
    </PropertyGroup>
```

**Comment by justinmchase** [comment]
```
dotnet_diagnostic.CS8619.severity = error
```

Shows up as an error in the IDE but is not reported as an error during the build, even with `GenerateDocumentationFile` set to true.

![Screen Shot 2022-04-07 at 10 03 39 PM](https://user-images.githubusercontent.com/10974/162354809-31bb48ff-55ec-4d86-ba59-bdb505e3d1fa.png)

**Comment by jnm2** [comment]
@justinmchase I wouldn't expect GenerateDocumentationFile to have anything to do with that diagnostic. On a side note, does setting https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#enforcecodestyleinbuild in your csproj help?

**Comment by justinmchase** [comment]
```xml
<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```

Does _not_ cause build errors when the above rule is set to `severity = error`

**Comment by osmnorhn** [comment]
I am also interested in status of this issue.

**Comment by ismaelhamed** [comment]
For me, setting both `EnforceCodeStyleInBuild` and `GenerateDocumentationFile` finally made the compilation failed for `IDE0005` but not for `CS8019`. 

```
dotnet_diagnostic.CS8019.severity = error
dotnet_diagnostic.IDE0005.severity = error
```

**Comment by Youssef1313** [comment]
@ismaelhamed I think you should never need CS8019

**Comment by ismaelhamed** [comment]
@justinmchase what's the point of `CS8019` then?

**Comment by Youssef1313** [comment]
@ismaelhamed CS8019 is what makes IDE0005 work. The idea is that it's harder to implement "unused usings" analyzer outside of the compiler. So, this is implemented as a hidden diagnostic (CS8019) in the compiler, and IDE0005 just bridges CS8019 to the user.

https://github.com/dotnet/roslyn/blob/28754bb176aa35823cb1cca60b259448c856d0a8/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Helpers/RemoveUnnecessaryImports/CSharpUnnecessaryImportsProvider.cs#L23-L44

https://github.com/dotnet/roslyn/blob/28754bb176aa35823cb1cca60b259448c856d0a8/src/Analyzers/Core/Analyzers/RemoveUnnecessaryImports/AbstractRemoveUnnecessaryImportsDiagnosticAnalyzer.cs#L65-L86

**Comment by ismaelhamed** [comment]
@Youssef1313 Got it now, thanks! Key point is `CS8019` being "a hidden diagnostic".

**Comment by sharwell** [comment]
The IDE also uses CS8019 to fade out unnecessary usings in generated code files, where it would never actually report an IDE0005 diagnostic. CS8019 reports _all_ unnecessary usings, while IDE0005 filters those down to only report actionable ones.

**Comment by shuebner** [comment]
TL;DR for the lazy people:
```xml
<PackageReference Include="SvSoft.MSBuild.CheckUnnecessaryUsings" Version="1.0.0" PrivateAssets="all" />
```

@jnm2's [workaround](https://github.com/dotnet/roslyn/issues/41640#issuecomment-985780130) triggered me to write my own (different) solution as a nuget package that enables IDE0005 warnings in the CLI build.
It works as expected with both documentation generation enabled and disabled.
No manual intervention necessary.
See my repo for more information: https://github.com/shuebner/CheckUnnecessaryUsings

Bug reports are of course welcome :-)

**Comment by glen-84** [comment]
@mavasani

This is now a warning in `7.0.400`:

> CSC : warning EnableGenerateDocumentationFile: Set MSBuild property 'GenerateDocumentationFile' to 'true' in project file to enable IDE0005 (Remove unnecessary usings/imports) on build (https://github.com/dotnet/roslyn/issues/41640)

This issue is now 3 and a half years old – when will it be resolved?

**Comment by VincentH-Net** [comment]
> Just FYI: This will produce warnings in builds that previously didn't have warnings. If your CI treats warning as errors, you will now have a build that mysteriously fails when it used to build without error.
> 
> It's unavoidable, but I'm just leaving this explanation for future readers who also pushed changes and wondered how those changes could _possibly_ have led to a compile error of this kind.

For my projects that do not need XML doc comments I prefer to disable this new warning:
```xml
<NoWarn>EnableGenerateDocumentationFile</NoWarn> <!-- Remove this to get a warning how to enable IDE0005 (Remove unnecessary usings/imports) on build; that warning describes a workaround for https://github.com/dotnet/roslyn/issues/41640 -->
```
Very annoying to have to modify all my project files as a workaround to a workaround warning!

Why is it so important to have IDE0005 (Remove unnecessary usings/imports) enabled? 
Does _not_ removing unnecessary usings/imports cause with a runtime / build penalty? 
Because enabling XML comments like the warning says _will_ make the build slower... is that worth it?

**Comment by sharwell** [comment]
@VincentH-Net Disabling `EnableGenerateDocumentationFile` will not work around this issue. In your project, IDE0005 has been _explicitly_ set to either warning or error, yet the build is unable to satisfy the request. This is not a default configuration, but rather one that has been customized as part of your build. If you do not wish to produce documentation files, the solution to this issue is to reduce the severity of IDE0005.

**Comment by sharwell** [comment]
> Just FYI: This will produce warnings in builds that previously didn't have warnings.

@mvonballmo This is correct. In all cases where new warnings have appeared, users have explicitly requested these warnings to appear, but prior versions of the compiler silently failed to report them.

**Comment by mvonballmo** [comment]
> > Just FYI: This will produce warnings in builds that previously didn't have warnings.
> 
> @mvonballmo This is correct. In all cases where new warnings have appeared, users have explicitly requested these warnings to appear, but prior versions of the compiler silently failed to report them.

I appreciate your taking the time to clarify that the warning appears because _a solution's configuration explicitly requests it_.

I hope others have understood that the only reason a build used to pass without warning (or error when warnings-as-errors is enabled) is because the checker was not actually fulfilling the contract requested in the configuration.

Anyone who wants neither a warning nor documentation-generation should just reduce the severity of that rule.

**Comment by pziezio** [comment]
The solution I'm using in a large project with CI pipelines is as follows:

Since I want all specified warnings and rules to be checked (IDE0005 included) I set `<GenerateDocumentationFile>true</GenerateDocumentationFile>` globally in the `Directory.Build.props` file.
One such file in the root folder applies to all projects in the solution.
There is an obvious downside in greater compilation time required to produce all XML documentation files.

In my CI pipeline all feature branches build the solution with these checks enabled.
My release (main) branch is published with code style checks disabled:
`dotnet publish ... /p:GenerateDocumentationFile=False /p:EnforceCodeStyleInBuild=False`

Since the only way to merge code into the main branch is through the feature branches, I'm sure that the merged code passed all style checks before merging. Furthermore, the published project does not contain any XML documentation files.

**Comment by sharwell** [comment]
@RehanSaeed @lonix1 This is already described in detail in my comment above, including all the options to proceed:
https://github.com/dotnet/roslyn/issues/41640#issuecomment-1699278994

**Comment by sharwell** [comment]
> Assuming I understand, I can no longer ... but the result is it's different to how it was before

The core compiler behavior did not change. The _only_ change is that a warning is now alerting you to the fact that the compiler was (always was, and still is) ignoring your requested configuration.

> But my builds are definitely noisy now.

The description of the new diagnostic contains information about how to eliminate the noise. It's generally as simple as this:
https://github.com/dotnet/roslyn-analyzers/blob/8281feab05b88e0fe99e0b9614b39a3ac79eeb58/Directory.Build.props#L34-L46

**Comment by RehanSaeed** [comment]
@sharwell Thanks for the explanation. From a developer experience point of view, it'd be nicer if we didn't have to disable IDE0005 in our .editorconfig files. If I now do so and in the future Roslyn decides to fix this issue, I will have IDE0005 disabled with no way to know that this bug has been fixed.

**Comment by sharwell** [comment]
@RehanSaeed The easier solution is to just enable documentation files in all projects. If you disable the warnings about missing documentation like I linked above, there is almost no downside to this (I've seen reports of bugs in unrelated tools not knowing how to handle these files, but those tools are outside our control in case that impacts you).

**Comment by VictorioBerra** [comment]
I am struggling a bit. When I set to `<GenerateDocumentationFile>true</GenerateDocumentationFile>`  I still get CS1573,CS1591,CS1712 errors even if I do:

`<NoWarn Condition="'$(Language)' == 'C#'">$(NoWarn),1573,1591,1712</NoWarn>`.

The only thing that fixes it for me without toggling `GenerateDocumentationFile` is just `<NoWarn>IDE0005</NoWarn>`.

What am I doing wrong?

**Comment by VictorioBerra** [comment]
So this works:

`<NoWarn>$(NoWarn),CS1591</NoWarn>`

But this doesnt:

`<NoWarn Condition="'$(Language)' == 'C#'">$(NoWarn),1573,1591,1712</NoWarn>`

**Comment by sharwell** [comment]
@VictorioBerra I'm guessing you have that line earlier in the build than wherever `$(Language)` is defined.

**Comment by smkanadl** [comment]
@VictorioBerra $(Language) is not supported in a Directory.Build.props but only in a Directory.Build.targets file.

**Comment by jnm2** [comment]
If I remember right, I've used `$(MSBuildProjectExtension)` for this purpose.

**Comment by AraHaan** [comment]
I don't like this one but because I do not want things like documentation files to be output when:

I build an winforms project where everything is "internal" visibility
  - has dlls with apis that can be used to load a "plugin interface".
  - said plugin interface is then used to extend the winforms project to add new features to it (although limited) on what can be added to it.
    - plugins themselves are forced to not generate documentation files as well by an msbuild sdk they are then "required" to bring in for application plugin development.

And yes I have an application of this type and the csc warning from doing this annoys me a lot.

**Comment by sharwell** [comment]
> It seems .editorconfig cannot be used to silence this warning. This is unacceptable.

This is intentional, similar to what would happen if you add the following to your project:

```xml
<ItemGroup>
  <Compile Include="ThisFileDoesNotExist.cs"/>
</ItemGroup>
```

By explicitly providing the compiler with an instruction to enforce IDE0005 at warning (or greater) severity, and then removing its ability to do so by disabling the documentation file output, the project is in an invalid state. Multiple solutions are provided above to return the project to a valid state (either disable the IDE0005 analysis, or enable the documentation output).

**Comment by AraHaan** [comment]
> > It seems .editorconfig cannot be used to silence this warning. This is unacceptable.
> 
> This is intentional, similar to what would happen if you add the following to your project:
> 
> ```
> <ItemGroup>
>   <Compile Include="ThisFileDoesNotExist.cs"/>
> </ItemGroup>
> ```
> 
> By explicitly providing the compiler with an instruction to enforce IDE0005 at warning (or greater) severity, and then removing its ability to do so by disabling the documentation file output, the project is in an invalid state. Multiple solutions are provided above to return the project to a valid state (either disable the IDE0005 analysis by setting the severity to suggestion or lower, or enable the documentation output).

Which is funny because I never really enabled that analysis on it specifically for outside of Visual Studio 😄.

**Comment by sharwell** [comment]
> Which is funny because I never really enabled that analysis on it specifically for outside of Visual Studio 😄.

IDE0005 is never enabled at warning severity or greater by default. There are many ways a project could be modified to increase its severity, so without a binlog its difficult to be more specific.

**Comment by mwasson74** [comment]
VS 17.9.2 and .NET 8 project and with all of the extra things added, the build is still not failing. Am I missing something?

![image](https://github.com/dotnet/roslyn/assets/12786751/16a7c22b-0226-470f-aba5-cde86894c1e9)

**Comment by RenderMichael** [comment]
Hi, I'm being affected by this:
![image](https://github.com/dotnet/roslyn/assets/12619902/5dcc0f35-cc7b-4068-bd14-da81d826ced5)

More specifically, I am generating the documentation file on every project except the testing projects. In my `Directory.Build.props` file I have

```xml
<GenerateDocumentationFile>true</GenerateDocumentationFile>
```

Directly in the tests project I then set

```xml
<GenerateDocumentationFile>false</GenerateDocumentationFile>
```

Currently I can get around this by `NoWarn`ing on `EnableGenerateDocumentationFile`, but would it be possible to ignore this for projects where `GenerateDocumentationFile` is explicitly set to `false`?

**Comment by sharwell** [comment]
@RenderMichael instead of disabling `EnableGenerateDocumentationFile`, you should either enable documentation output in your test projects or disable `IDE0005`. As it stands, your test projects have directly instructed the compiler to perform an operation which is not supported, and it is resulting in a warning.

**Comment by RenderMichael** [comment]
Thanks for the response.

It's not clear from the user side that IDE0005 requires documentation files to be generated. Can the error message include something along the lines of `IDE0005 requires documentation files to be generated. Please enable or disable both in this project`?

**Comment by rampaa** [comment]
I have the following in my `Directory.build.props` file:

```xml
<GenerateDocumentationFile>true</GenerateDocumentationFile>
<CopyDocumentationFileToOutputDirectory>false</CopyDocumentationFileToOutputDirectory>
<PublishSingleFile>true</PublishSingleFile>
```

This results in the following error:

```
  JL.Windows net9.0-windows win-x64 failed with 1 error(s) (0.4s) → C:\Users\User\source\repos\JL\JL.Windows\bin\Release\net9.0-windows\win-x64\JL.dll
    C:\Program Files\dotnet\sdk\10.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(1132,5): error MSB4018:
      The "GenerateBundle" task failed unexpectedly.
      System.IO.FileNotFoundException: Could not find file 'C:\Users\User\source\repos\JL\JL.Windows\bin\Release\net9.0-
      windows\win-x64\JL.xml'.
      File name: 'C:\Users\User\source\repos\JL\JL.Windows\bin\Release\net9.0-windows\win-x64\JL.xml'
         at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, Fil
      eShare share, FileOptions options)
         at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare
       share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
         at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare sha
      re, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
         at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, Fil
      eShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode)
         at System.IO.File.OpenRead(String path)
         at Microsoft.NET.HostModel.Bundle.Bundler.IsAssembly(String path, Boolean& isPE)
         at Microsoft.NET.HostModel.Bundle.Bundler.InferType(FileSpec fileSpec)
         at Microsoft.NET.HostModel.Bundle.Bundler.GetFilteredFileSpecs(IEnumerable`1 fileSpecs)
         at Microsoft.NET.HostModel.Bundle.Bundler.GenerateBundle(IReadOnlyList`1 fileSpecs)
         at Microsoft.NET.Build.Tasks.GenerateBundle.<>c__DisplayClass56_0.<ExecuteWithRetry>b__0()
         at Microsoft.NET.Build.Tasks.GenerateBundle.DoWithRetry(Action action)
         at Microsoft.NET.Build.Tasks.GenerateBundle.ExecuteWithRetry()
         at Microsoft.NET.Build.Tasks.GenerateBundle.ExecuteCore()
         at Microsoft.NET.Build.Tasks.TaskBase.Execute()
         at Microsoft.Build.BackEnd.TaskExecutionHost.Execute()
         at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(TaskExecutionHost taskExecutionHost, TaskLogging
      Context taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)

Build failed with 1 error(s) in 0.7s
```

Is this the expected behavior or should I open an issue about this?

**Comment by keystroke** [comment]
any update on this? just ran into this trying to add style and lint rules to help with AI agent workflows and enforcing the usings thing would be nice, but enabling xml doc generation is annoying and not worth it :/
